src/Middleware/ChaosHandler.cs

```diff
@@ -1,256 +1,264 @@
-            public string Message { get; set; }
-        }
-    }
-}
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Kiota.Http.HttpClientLibrary.Extensions;
+using Microsoft.Kiota.Http.HttpClientLibrary.Middleware.Options;
+
+namespace Microsoft.Kiota.Http.HttpClientLibrary.Middleware
+{
+    /// <summary>
+    /// A <see cref="DelegatingHandler"/> implementation that is used for simulating server failures.
+    /// </summary>
+    public class ChaosHandler : DelegatingHandler, IDisposable
+    {
+        private readonly Random _random;
+        private readonly ChaosHandlerOption _chaosHandlerOptions;
+        private List<HttpResponseMessage> _knownFailures;
+        private const string Json = "application/json";
+
+        /// <summary>
+        /// Create a ChaosHandler.
+        /// </summary>
+        /// <param name="chaosHandlerOptions">Optional parameter to change default behavior of handler.</param>
+        public ChaosHandler(ChaosHandlerOption chaosHandlerOptions = null)
+        {
+            _chaosHandlerOptions = chaosHandlerOptions ?? new ChaosHandlerOption();
+            _random = new Random(DateTime.Now.Millisecond);
+            LoadKnownFailures(_chaosHandlerOptions.KnownChaos);
+        }
+        /// <summary>
+        /// The key used for the open telemetry event.
+        /// </summary>
+        public const string ChaosHandlerTriggeredEventKey = "com.microsoft.kiota.chaos_handler_triggered";
+        /// <summary>
+        /// Sends the request
+        /// </summary>
+        /// <param name="request">The request to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request.</param>
+        /// <returns></returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if(request == null)
+                throw new ArgumentNullException(nameof(request));
+
+            // Select global or per request options
+            var chaosHandlerOptions = request.GetRequestOption<ChaosHandlerOption>() ?? _chaosHandlerOptions;
+            ActivitySource activitySource;
+            Activity activity;
+            if (request.GetRequestOption<ObservabilityOptions>() is ObservabilityOptions obsOptions) {
+                activitySource = new ActivitySource(obsOptions.TracerInstrumentationName);

```

Comment:
Similar comment to https://github.com/microsoft/kiota-authentication-azure-dotnet/pull/28#discussion_r975093205